### PR TITLE
Removing Nunit from utils

### PR DESCRIPTION
**Pull request by psfinaki**
Related: https://github.com/dotnet/fsharp/issues/13654

TL;DR: We are nearly done!
- This removes Nunit from test frameworks
- Cleans things up
- **Enables 32 valid ComponentTests** that were silently not being executed

Please see my comments everywhere, this is close to the equivalent code, with minor simplifications and fixes.

---

Note this increases _the number_ of tests executed in the CI. The execution time is not affected.

Main:
image

This branch:
image

This is due to:
- 5 new tests in the `FSharpSuite`. Those are old-new `AsyncTests` and they are skipped anyway for now. See comment.
- 35 new tests in the `ComponentTests.TypeChecks`. Of those:
    - 2  `CompilationFromCmdlineArgsTests` and 1 `TypedTreeGraphTests` -  those were explicit + failing, now skipped
    - 1 `DependencyResolutionTests` - this is **legit and passing**, wasn't being executed as `[<Theory>]` was forgotten there
    - 8 `FileContentMappingTests`, 1 `GraphProcessingTests`, 4 `QueryTrieTests`, 18 `TrieMappingTests` - those are **legit and passing**, weren't being executed because weren't rewritten to Xunit (so trying executing non-existing Nunit adapter)

### CODE REVIEW

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/CompilationFromCmdlineArgsTests.fs [hunk 1]
```diff
@@ -5,19 +5,19 @@ module CompilationFromCmdlineArgsTests =
...
     // The path needs to be absolute.
-    let localProjects: string list =
+    let localProjects =
         [
             @"C:\Projects\fantomas\src\Fantomas.Core\Fantomas.Core.args.txt"
             @"C:\Projects\FsAutoComplete\src\FsAutoComplete\FsAutoComplete.args.txt"
             @"C:\Projects\fsharp\src\Compiler\FSharp.Compiler.Service.args.txt"
             @"C:\Projects\fsharp\tests\FSharp.Compiler.ComponentTests\FSharp.Compiler.ComponentTests.args.txt"
-        ]
+        ] |> Seq.map (fun p -> [| box p |])
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/CompilationFromCmdlineArgsTests.fs:20]: This is Xunit specific, `MemberData` must be `obj array list` or `obj array array`. We do this in all similar places, e.g. [here](https://github.com/dotnet/fsharp/blob/398d16d95c14c96f5ff7173b5fa529fa879cbee6/tests/FSharp.Compiler.ComponentTests/Conformance/BasicGrammarElements/StaticLet/StaticLetInUnionsAndRecords.fs#L13).

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/CompilationFromCmdlineArgsTests.fs [hunk 2]
```diff
@@ -43,16 +43,16 @@ module CompilationFromCmdlineArgsTests =
...
-    [<Explicit("Slow, only useful as a sanity check that the test codebase is sound and type-checks using the old method")>]
+    [<MemberData(nameof localProjects)>]
+    [<Theory(Skip = "Slow, only useful as a sanity check that the test codebase is sound and type-checks using the old method")>]
     let ``Test sequential type-checking`` (projectArgumentsFilePath: string) =
         testCompilerFromArgs Method.Sequential projectArgumentsFilePath
 
-    [<TestCaseSource(nameof localProjects)>]
-    [<Explicit("This only runs with the explicitly mentioned projects above")>]
+    [<MemberData(nameof localProjects)>]
+    [<Theory(Skip = "This should only run with the explicitly mentioned projects above")>]
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/CompilationFromCmdlineArgsTests.fs:56]: There is [no](https://stackoverflow.com/questions/37334245/can-you-mark-xunit-tests-as-explicit) `Explicit` attribute in Xunit (skipping the test by default, but running on-demand). It [should arrive](https://github.com/xunit/xunit/issues/2518) in Xunit 3.

Hence just marking as skipped. This anyway fails by default, without projects located on the disk.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/DependencyResolutionTests.fs [hunk 3]
```diff
@@ -1,11 +1,14 @@
 ﻿module TypeChecks.DependencyResolutionTests
 
 open TypeChecks.TestUtils
-open NUnit.Framework
+open Xunit
 open FSharp.Compiler.GraphChecking
 open Scenarios
 
-[<TestCaseSource(nameof scenarios)>]
+let scenarios = scenarios |> Seq.map (fun p -> [| box p |])
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/DependencyResolutionTests.fs:8]: See the comment above, conversion is needed for Xunit's `MemberData`.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/DependencyResolutionTests.fs [hunk 4]
```diff
@@ -18,4 +21,4 @@ let ``Supported scenario`` (scenario: Scenario) =
     for file in scenario.Files do
         let expectedDeps = file.ExpectedDependencies
         let actualDeps = set graph.[file.Index]
-        Assert.AreEqual(expectedDeps, actualDeps, $"Dependencies don't match for {System.IO.Path.GetFileName file.FileName}")
+        Assert.True((expectedDeps = actualDeps), $"Dependencies don't match for {System.IO.Path.GetFileName file.FileName}")
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/DependencyResolutionTests.fs:24]: There is no overload of `Assert.Equal` with a message, this one is rather useful, so keeping it in this shape.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/FileContentMappingTests.fs [hunk 5]
```diff
@@ -45,10 +45,10 @@ module X.Y.Z
 """
 
     match content with
-    | [ TopLevelNamespace "X.Y" [] ] -> Assert.Pass()
+    | [ TopLevelNamespace "X.Y" [] ] -> ()
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/FileContentMappingTests.fs:48]: [No](https://stackoverflow.com/a/14631983/3232646) `Assert.Pass()` in Xunit. Just returning from the method does the trick here.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/TrieMappingTests.fs [hunk 6]
```diff
@@ -49,22 +49,22 @@ type C = { CX: int; CY: int }
...
     let xNode = trie.Children.["X"]
-    Assert.AreEqual(1, xNode.Children.Count)
+    Assert.Equal(1, xNode.Children.Count)
     Assert.True(Seq.isEmpty xNode.Files)
 
     let yNode = xNode.Children["Y"]
-    Assert.AreEqual(2, yNode.Children.Count)
-    Assert.AreEqual(set [| 2 |], yNode.Files)
+    Assert.Equal(2, yNode.Children.Count)
+    Assert.Equal<Set<FileIndex>>(set [| 2 |], yNode.Files)
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/TrieMappingTests.fs:57]: Xunit asserts same types by default and has many implicit overloads for `Assert.Equal` - but not for sets, hence specifying the types explicitly.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/TypedTreeGraph.fs [hunk 7]
```diff
@@ -100,8 +100,8 @@ let graphFromTypedTree (checker: FSharpChecker) (projectOptions: FSharpProjectOp
         return files, graph
     }
 
-[<TestCaseSource(nameof localProjects)>]
-[<Explicit("Slow! Only useful as a sanity check that the test codebase is sound.")>]
+[<MemberData(nameof localProjects)>]
+[<Theory(Skip = "Slow! Only useful as a sanity check that the test codebase is sound.")>]
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/TypedTreeGraph.fs:104]: Ditto, no explicit attribute in Xunit yet, hence just skipping.

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Core/IntConversionsTestGenerator.fsx [hunk 8]
```diff
@@ -45,7 +45,7 @@ do
 // This file is automatically generated by IntConversionsTestGenerator.fsx    
 namespace FSharp.Core.UnitTests        
 open System                           
-open NUnit.Framework
+open XUnit
 open FSharp.Core.UnitTests.LibraryTestFx
```
Comment (psfinaki) [inline tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Core/IntConversionsTestGenerator.fsx:49]: This here is not test code, but test generation code. This file is `IntConversionsTestGenerator.fsx` and the actual tests are located in [IntConversionsGenerated](https://github.com/dotnet/fsharp/blob/85617ced7e1d5aac0dd132c1fa973d5587c52927/tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Core/IntConversionsGenerated.fs).

Things were initially Nunit here and were generating Nunit. But 4 years ago this [migrated to](https://github.com/dotnet/fsharp/pull/9992/files#diff-ffa136487f3031a2cc3d55cf756257c200afa7705ac7cbf6f36d70c1a0134e8e) to Xunit. The generated file was updated, the generator was not - hence doing it here.

File: tests/FSharp.Test.Utilities/Compiler.fs [hunk 9]
```diff
@@ -1744,7 +1743,7 @@ Actual:
         let private assertEvalOutput (selector: FsiValue -> 'T) (value: 'T) (result: CompilationResult) : CompilationResult =
             match result.RunOutput with
             | None -> failwith "Execution output is missing cannot check value."
-            | Some (EvalOutput (Ok (Some e))) -> Assert.AreEqual(value, (selector e))
+            | Some (EvalOutput (Ok (Some e))) -> Assert.Equal<'T>(value, (selector e))
```
Comment (psfinaki) [inline tests/FSharp.Test.Utilities/Compiler.fs:1746]: Ditto, Xunit compares types by default, needs to be explicit here.

File: tests/FSharp.Test.Utilities/CompilerAssert.fs [hunk 10]
```diff
@@ -771,32 +771,32 @@ Updated automatically, please check diffs in your pull request, changes must be
...
 
     static member ExecutionHasOutput(cmpl: Compilation, expectedOutput: string) =
-        CompilerAssert.Execute(cmpl, newProcess = true, onOutput = (fun output -> Assert.AreEqual(expectedOutput, output, sprintf "'%s' = '%s'" expectedOutput output)))  
+        CompilerAssert.Execute(cmpl, newProcess = true, onOutput = (fun output -> Assert.Equal(expectedOutput, output)))  
 
     static member Pass (source: string) =
         let parseResults, fileAnswer = checker.ParseAndCheckFileInProject("test.fs", 0, SourceText.ofString source, defaultProjectOptions TargetFramework.Current) |> Async.RunImmediate
 
-        Assert.IsEmpty(parseResults.Diagnostics, sprintf "Parse errors: %A" parseResults.Diagnostics)
+        Assert.Empty(parseResults.Diagnostics)
```
Comment (psfinaki) [inline tests/FSharp.Test.Utilities/CompilerAssert.fs:779]: Xunit should print structured output by default here (i.e., in case of empty collection, it should structure-print existing elements).

File: tests/fsharp/Compiler/Libraries/Core/Async/AsyncTests.fs
Comment (psfinaki) [inline tests/fsharp/Compiler/Libraries/Core/Async/AsyncTests.fs:1]: Another bad merge consequence here.

The tests (nunit then) were introduced in [this PR](https://github.com/dotnet/fsharp/pull/7234/files) 5 years ago. But just a day later, the file was accidentally removed from the project in [this merge](https://github.com/dotnet/fsharp/pull/7272/files#diff-f69a1e6058f88408ad6c792b2f558912d8e26ebc7abeead2d8fa395ae9b87030). Since then, the tests were not being executed.

It was easy to revive them, but they all crash the test host now, so I skipping them with a remark.

File: tests/fsharp/FSharpSuite.Tests.fsproj [hunk 11]
```diff
@@ -76,6 +76,7 @@
     <Compile Include="Compiler\Regressions\NullableOptionalRegressionTests.fs" />
     <Compile Include="Compiler\Regressions\IndexerRegressionTests.fs" />
     <Compile Include="Compiler\Regressions\ForInDoMutableRegressionTest.fs" />
+    <Compile Include="Compiler\Libraries\Core\Async\AsyncTests.fs" />
```
Comment (psfinaki) [inline tests/fsharp/FSharpSuite.Tests.fsproj:79]: See comment to the tests themselves.

File: tests/fsharp/nunitConf.fs
Comment (psfinaki) [inline tests/fsharp/nunitConf.fs:1]: This used to be a test helper back in a day but then it was eliminated [8 years ago](https://github.com/dotnet/fsharp/pull/1750/files#diff-3f15493d9b2e9f5294e80d23efeab6ddf13fd375c7aecabacb14d30a979aa88c) during some test simplifications. What replaced it then and what's used now is [TestFramework.fs](https://github.com/dotnet/fsharp/blob/main/tests/FSharp.Test.Utilities/TestFramework.fs).

This file was just forgotten to be removed from disk.

File: tests/service/FsUnit.fs [hunk 12]
```diff
@@ -35,32 +35,4 @@ let shouldPairwiseEqual (x: seq<_>) (y: seq<_>) =
     while ex.MoveNext() do countx <- countx + 1
     while ey.MoveNext() do county <- county + 1
     if countx <> county then        
-        Assert.Fail("Collections are of unequal lengths, expected length {0}, actual length is {1}.", countx, county)
-
-let notEqual x = NotConstraint(EqualConstraint(x))
```
Comment (psfinaki) [inline tests/service/FsUnit.fs:40]: Most of these were not used at all. A few were used in a couple of places, I replaced them with equivalent (e.g. used `Assert.True` instead of `True` constraint).

File: tests/service/FsUnit.fs [hunk 13]
```diff
@@ -1,23 +1,23 @@
 ﻿module FsUnit
 
 open System.Diagnostics
-open NUnit.Framework
-open NUnit.Framework.Constraints
+open Xunit
 
 [<DebuggerNonUserCode>]
-let should (f : 'a -> #Constraint) x (y : obj) =
```
Comment (psfinaki) [inline tests/service/FsUnit.fs:8]: This all is not needed anymore really - Xunit checks same types by default in its assertions.
`should` is only used with `equal` now.

I will likely remove this completely, don't want to do it here to not explode the diff.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/DependencyResolutionTests.fs [hunk 14]
```diff
@@ -1,11 +1,14 @@
...
 open TypeChecks.TestUtils
-open NUnit.Framework
+open Xunit
 open FSharp.Compiler.GraphChecking
 open Scenarios
 
-[<TestCaseSource(nameof scenarios)>]
+let scenarios = scenarios |> Seq.map (fun p -> [| box p |])
+
+[<Theory>]
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/DependencyResolutionTests.fs:10]: Note that attribute was forgotten here so tests were not executing - they are now.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/FileContentMappingTests.fs [hunk 15]
```diff
@@ -1,6 +1,6 @@
 ﻿module TypeChecks.FileContentMappingTests
 
-open NUnit.Framework
+open Xunit
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/FileContentMappingTests.fs:3]: These tests were not executed at all. They were not rewritten to Xunit whereas the project had xunit adapter set up via dependencies. Now they run.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/GraphProcessingTests.fs [hunk 16]
```diff
@@ -2,9 +2,9 @@
 
 open System.Threading
 open FSharp.Compiler.GraphChecking.GraphProcessing
-open NUnit.Framework
+open Xunit
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/GraphProcessingTests.fs:5]: Ditto, was not executed at all due to the test framework mixture.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/QueryTrieTests.fs [hunk 17]
```diff
@@ -2,7 +2,7 @@
 
 open System.Collections.Generic
 open System.Collections.Immutable
-open NUnit.Framework
+open Xunit
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/QueryTrieTests.fs:5]: Ditto, was not executed at all due to the test framework mixture.

File: tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/TrieMappingTests.fs [hunk 18]
```diff
@@ -1,14 +1,14 @@
 ﻿module TypeChecks.TrieMappingTests
 
-open NUnit.Framework
+open Xunit
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/TypeChecks/Graph/TrieMappingTests.fs:3]: Ditto, was not executed at all due to the test framework mixture.

**Review by T-Gro — APPROVED** [review]

**Review by abonie — APPROVED** [review]
